relayer/client.go

```diff
@@ -389,3 +392,46 @@ func MustGetHeight(h ibcexported.Height) clienttypes.Height {
 	}
 	return height
 }
+
+// FindMatchingClient is a helper function that will determine if there exists a client with identical client and
+// consensus states to the client which would have been created. Source is the chain that would be adding a client
+// which would track the counterparty. Therefore, we query source for the existing clients
+// and check if any match the counterparty. The counterparty must have a matching consensus state
+// to the latest consensus state of a potential match. The provided client state is the client
+// state that will be created if there exist no matches.
+func FindMatchingClient(ctx context.Context, src, dst *Chain, newClientState ibcexported.ClientState) (string, bool, error) {
+	var (
+		clientsResp clienttypes.IdentifiedClientStates
+		err         error
+	)
+
+	if err = retry.Do(func() error {
+		clientsResp, err = src.ChainProvider.QueryClients(ctx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}, retry.Context(ctx), RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
+		src.log.Info(
+			"Failed to query clients",
+			zap.String("chain_id", src.ChainID()),
+			zap.Uint("attempt", n+1),
+			zap.Uint("max_attempts", RtyAttNum),
+			zap.Error(err),
+		)
+	})); err != nil {
+		return "", false, err
+	}
+
+	for _, existingClientState := range clientsResp {
+		clientID, found, err := provider.ClientsMatch(ctx, src.ChainProvider, dst.ChainProvider, existingClientState, newClientState)
+		if err != nil {
+			return "", false, err
```

Comment:
Since this is in a loop, it might be better to just log the error, so that one client which fails to parse doesn't break everything else.

Then again, what if we fail to parse the one that should match? Maybe just a comment here would clarify the intent.

Reply:
I leaned more towards leaving the returned error and adding a comment explaining that realistically we should not be encountering an error here and if we do it is likely a sign that there are bigger issues at play.